core/src/main/java/nl/nn/adapterframework/util/CompactSaxHandler.java

```diff
@@ -94,64 +91,86 @@ public void endElement(String uri, String localName, String qName) throws SAXExc
 			throw new SAXException("expected end element [" + lastElement + "] but got end element [" + localName + "]");
 		}
 
-		printCharData();
+		printCharData(false);
 		if (isRemoveCompactMsgNamespaces()) {
-			messageBuilder.append("</").append(localName).append(">");
+			super.endElement(uri, localName, localName);
 		} else {
-			messageBuilder.append("</").append(qName).append(">");
+			super.endElement(uri, localName, qName);
 		}
 		elements.remove(lastIndex);
 	}
 
 	@Override
-	public void characters(char[] ch, int start, int length) throws SAXException {
+	public void characters(char[] ch, int start, int length) {
 		charDataBuilder.append(ch, start, length);
 	}
 
-	private void printCharData() {
+	private void printCharData(boolean startElement) throws SAXException {
 		if (charDataBuilder.length() == 0) {
 			return;
 		}
 
-		String before = "";
-		String after = "";
-
-		if (chompLength >= 0 && charDataBuilder.length() > chompLength) {
-			before = "*** character data size [" + charDataBuilder.length() + "] exceeds [" + getChompCharSize() + "] and is chomped ***";
-			after = "...(" + (charDataBuilder.length() - chompLength) + " characters more)";
-			charDataBuilder.setLength(chompLength);
+		// Detection of MOVE_START and MOVE_END session key fragments, on end-Element phase
+		int length = charDataBuilder.length();
+		if (!startElement
+				&& length > VALUE_MOVE_START.length()
+				&& charDataBuilder.substring(length - 1, length).equals(VALUE_MOVE_END)
+				&& charDataBuilder.substring(0, VALUE_MOVE_START.length()).equals(VALUE_MOVE_START)) {
```

Comment:
These checks are quite expensive and only need to be done if `moveElementFound == true && session != null`, can be changed to a direct assignment:
```
moveElementFound = moveElementFound && session != null && !startElement && length > VALUE_MOVE_START.length()
 				&& charDataBuilder.substring(length - 1, length).equals(VALUE_MOVE_END)
 				&& charDataBuilder.substring(0, VALUE_MOVE_START.length()).equals(VALUE_MOVE_START)
```

Reply:
I've changed the logic as you suggested, but with an important negation added.

I did also do a benchmark between the previous version (before applying these suggestions). 
The results are not as positive as what you've hoped for: 
100KB XML: 2,2% decrease in the first test, 0,25% decrease in the second test.
1MB XML: 0,8% decrease in the first test, 0,8% increase in the second test. 
Perhaps statistically not relevant, since it is lower then the error margin.